src/sentry/eventstream/snuba.py

```diff
@@ -157,32 +160,37 @@ def insert(
 
         is_transaction_event = self._is_transaction_event(event)
 
+        extra_data = [
+            {
+                "group_id": event.group_id,
+                "group_ids": [group.id for group in event.groups],
+                "event_id": event.event_id,
+                "organization_id": project.organization_id,
+                "project_id": event.project_id,
+                # TODO(mitsuhiko): We do not want to send this incorrect
+                # message but this is what snuba needs at the moment.
+                "message": event.search_message,
+                "platform": event.platform,
+                "datetime": event.datetime,
+                "data": event_data,
+                "primary_hash": primary_hash,
+                "retention_days": retention_days,
+            },
+            {
+                "is_new": is_new,
+                "is_regression": is_regression,
+                "is_new_group_environment": is_new_group_environment,
+                "skip_consume": skip_consume,
+            },
+        ]
+
+        if groups_state is not None:
+            extra_data.append({**groups_state})
```

Comment:
Hmm, maybe I'm not totally understanding the `EventStream` inheritance hierarchy. It looks like the `SnubaEventStream` doesn't make use of the extra data, but `KafkaEventStream` does? 

Then the `PostProcessForwarderWorker` processes the inserted payload (including message and headers) from kafka, which attempts to parse from the headers for the appropriate task kwargs and but will fallback to the message body if the header data is invalid. 

https://github.com/getsentry/sentry/blob/e126fbee41ea50543f55df64deaeea969531e517/src/sentry/eventstream/kafka/postprocessworker.py#L43-L56

For completeness, it might makes sense to include the group_state data in the header and message but I don't mind removing the extra data in the message itself.

Reply:
`SnuabEventStream` makes use of the extra data here: https://github.com/getsentry/sentry/blob/7e85e5de95a9604cfaa81339448ee9db2f4a083d/src/sentry/eventstream/snuba.py#L393

You're right that `KafkaEventStream` uses it here: https://github.com/getsentry/sentry/blob/7e85e5de95a9604cfaa81339448ee9db2f4a083d/src/sentry/eventstream/kafka/backend.py#L186. But that's just for writing to the message body. These params are really only useful for the post process forwarder, snuba doesn't care about them. I think it's actually safe to not modify the message body, that way we don't have any concerns about breaking anything in the snuba consumer.

Post process forwarder actually only uses the headers. For the sake of understanding, assume the `post-process-forwarder:kafka-headers` and `eventstream:kafka-headers` options are always true. The exception path was just there to handle the cutover and as a failsafe. I don't see any occurrences of the error there in sentry in the last 90 days.

The old paths that process messages are no longer used and haven't been for a while now. I'd actually like to remove them as part of this work (separate pr though).

Reply:
The header hack was needed until we find a better long term solution for scaling the post process forwarder. It's the header option that was originally planned to be removed, not the other way around.

Reply:
I didn't realise that the header thing was temporary. Tbh it seems like a good solution. Even if we scale this better using arroyo/remove semantic partitioning/etc, this results these consumers doing a lot less work overall

Reply:
Fair enough. I'll assume that submitting the extra GroupStates to the header should be enough and remove the extra params from the body.

Reply:
We had many long discussions about this in the past. Ultimately the team (search and storage) agreed that the header can considered an optimization but can NOT be entirely relied on for important data which must also be in the body as well. I'm not sure if we want to revisit this decision but I think I'd prefer to keep it the same way right now and put it in both places.

Reply:
Should I include the extra data in the message as well? Right now, its only in the header.